compiler/verif/z3refinement.ml

```diff
@@ -1721,23 +1742,30 @@ and vc_gen_typ_exp_desc ctx env typenv t =
 
       Creates z3 vc_gen_expression from type vc_gen_expression and adds it to the environment
 *)
-  match t.desc with
-  | Etypevar(n) -> debug(Printf.sprintf "Etypevar %s\n" n)
-  | Etypeconstr(t, txp_list) -> debug(Printf.sprintf "Etypeconstr\n"); qualident t; (List.iter (vc_gen_typ_exp_desc ctx env typenv) txp_list) 
-  | Etypetuple(txp_list) -> debug(Printf.sprintf "Etypetuple\n"); (List.iter (vc_gen_typ_exp_desc ctx env typenv) txp_list)
-  | Etypevec(texp , si) -> debug(Printf.sprintf "Etypevec\n")
-  | Etypefun(k, t, texp, texp2) -> debug(Printf.sprintf "Etypefun\n")
-  | Etypefunrefinement(k, t, te, te2, e) -> debug(Printf.sprintf "Etypefunrefinement\n")
-  | Erefinement(t, e) -> debug(Printf.sprintf "Erefinement\n");  
-       let expr = (vc_gen_expression ctx env e typenv) in
-       (debug(Printf.sprintf "Returning from e local: %s\n" (Expr.to_string expr));
-       (debug(Printf.sprintf "t.name %s" (fst t)));
-       (* add_constraint env expr; *)
-       z3_solve ctx env expr;
-       )
-  | Erefinementpairfuntype(txp_list, exp) -> debug(Printf.sprintf "Erefinementfunpair \n")
-       (* List.iter (fun elem ->         ) txp_list *)
 
+match t.desc with
+| Etypevar(n) -> debug(Printf.sprintf "Etypevar %s\n" n)
+| Etypeconstr(t, txp_list) -> 
+  debug(Printf.sprintf "Etypeconstr\n"); qualident t; 
+  List.iter (fun x -> vc_gen_typ_exp_desc ctx env typenv x None) txp_list
+  | Etypetuple(txp_list) -> 
+    debug(Printf.sprintf "Etypetuple\n"); 
+    List.iter (fun x -> vc_gen_typ_exp_desc ctx env typenv x None) txp_list
+    | Etypevec(texp , si) -> debug(Printf.sprintf "Etypevec\n")
+    | Etypefun(k, t, texp, texp2) -> debug(Printf.sprintf "Etypefun\n")
+    | Etypefunrefinement(k, t, te, te2, e) -> debug(Printf.sprintf "Etypefunrefinement\n")
+    | Erefinement(t, e) -> 
+      debug(Printf.sprintf "Erefinement\n");
+      let expr = vc_gen_expression ctx env e typenv in
+      (* get the variable name from the constriant and replace them by the reference variable *)
```

Comment:
I think it might just be a terminology thing, but the code seems to replace the refinement variable with a variable from the function body, rather than a reference variable